PHP-189 - Map compare doesn't compare values

mpenick opened the pull request:
`php_driver_map_compare()` doesn't compare values so if two maps have
the same keys, but different values they will compare as being equal.

Commit:
PHP-189 - Map compare doesn't compare values

`php_driver_map_compare()` doesn't compare values so if two maps have
the same keys, but different values they will compare as being equal.

stamhankar999 commented on ext/src/Map.c:
```diff
@@ -518,11 +518,14 @@ php_driver_map_compare(zval *obj1, zval *obj2 TSRMLS_DC)
   }
```
This clause has a typo; it should be comparing map1 to map2!

mpenick replied:
Good catch!

stamhankar999 commented on ext/src/Map.c:
```diff
@@ -518,11 +518,14 @@ php_driver_map_compare(zval *obj1, zval *obj2 TSRMLS_DC)
     HASH_FIND_ZVAL(map2->entries, PHP5TO7_ZVAL_MAYBE_P(curr->key), entry);
     if (entry == NULL) {
       return 1;
     }
```
It makes me a little nervous that there's no circumstance where we'll return -1 for a key in map2 that is missing in map1. Also, does HASH_ITER iterate the map entries in a deterministic way? I think this compare function does a great job of distinguishing equality/inequality. But I don't think it would be safe to be used for sorting a collection of maps (which may be more a theoretical concern than practical, but you never know).

Example:
map1 has one entry: k1, v1
map2 has one entry: k2, v2

php_driver_map_compare(map1, map2) => 1
php_driver_map_compare(map2, map1) => 1

I think in php-land that translates to `map1 < map2` and `map2 < map1` both being true, which make maps ill-suited to be stored in sorted collections (e.g. a binary search of such a collection may not find a map).

Upon further reflection, a key existing in map1 that does *not* exist in map2 implies that there is a key in map2 that doesn't exist in map1, since map sizes are verified to be the same by the time we reach this part of the code. So then the question boils down to which "missing key" is greater/less than the other "missing key" to produce a stable ordering/comparison of the maps. `php_driver_value_compare`s of missing keys would resolve that.

It sounds expensive, but it would be more correct.

mpenick replied:
Yeah, I thought about this when originally implementing it and favored the fast and easy to implement route. I think that documenting the limitation might be a better route here. What do you think?

mpenick replied:
> Also, does HASH_ITER iterate the map entries in a deterministic way?

No, it uses insertion order.

stamhankar999 replied:
Yeah, let's just document it.

mpenick replied:
For better or worse. I just discovered that PHP's internal hash table does the same thing for arrays with keys: https://github.com/php/php-src/blob/master/Zend/zend_hash.c#L2352-L2364

```php
array("a" => 1) <=> array("b" => 1);
# => 1
array("b" => 1) <=> array("a" => 1);
# => 1
```

stamhankar999 replied:
Very interesting.

Commit:
PHP-189 - Map compare doesn't compare values

Fixed typo in Map and Set that would cause collections of different
counts to compare incorrectly.

mpenick commented:
Merged at 63935137c3d7a873d9443335345bd718b2454379
